File: trunk/Demos/Crazy Stars/MenuScene.cpp
#include "StdAfx.h"
#include "MenuScene.h"
#include "MenuItem.h"

CMenuScene::CMenuScene()
	: m_pApp(GetRUGE())
	, m_pscStars(NULL)
	, m_nLastID(0)
	, m_hTex(NULL)
	, m_hSound(NULL)
	, m_hFont(NULL)
	, m_pGUI(NULL)
	, m_pSpr(NULL)
{
	m_pscStars=new CStarsScene(this);
}

CMenuScene::~CMenuScene()
{
	delete m_pscStars;
	m_pApp->Release();
}

BOOL CMenuScene::EnterScene(WPARAM wParam, LPARAM lParam)
{
	m_hTex=m_pApp->Texture_Load("cursor.png");
	m_hSound=m_pApp->Effect_Load("menu.wav");
	m_hFont=m_pApp->Font_Create(40, 0, 0, FALSE, "Î¢ÈíÑÅºÚ");

	m_pSpr=new CSprite(m_hTex, 0, 0, 32, 32);

	m_pGUI=new CGUI;
	m_pGUI->AddCtrl(new CMenuItem(1, m_hFont, m_hSound, 400, 240, 0.1f, "Start"));
	m_pGUI->AddCtrl(new CMenuItem(2, m_hFont, m_hSound, 400, 280, 0.2f, "Resume"));
	m_pGUI->AddCtrl(new CMenuItem(3, m_hFont, m_hSound, 400, 320, 0.3f, "Exit"));
	m_pGUI->SetNavMode(NAV_UPDOWN|NAV_CYCLED);
	m_pGUI->SetCursor(m_pSpr);
	m_pGUI->SetFocus(1);
	m_pGUI->Enter();

	m_nLastID=0;

	return TRUE;
}

void CMenuScene::ExitScene()
{
	delete m_pGUI;
	delete m_pSpr;
	m_pApp->Font_Free(m_hFont);
	m_pApp->Audio_Free(m_hSound);
	m_pApp->Texture_Free(m_hTex);
}

BOOL CMenuScene::Update(CSceneManager *pSceneManager, float fDelta)
{
	if (m_pApp->Input_GetKey()==VK_ESCAPE)
	{
		m_nLastID=3;
		m_pGUI->Exit();
	}

	int nID=m_pGUI->Update(fDelta);

	if (nID==-1)
	{
		switch (m_nLastID)
		{
		case 1:
			pSceneManager->SwitchScene(m_pscStars, ENTER_START);
			return FALSE;
		case 2:
			pSceneManager->SwitchScene(m_pscStars, ENTER_RESUME);
			return FALSE;
		case 3:
			return TRUE;
		}
	}
	else if (nID>0)
	{
		m_nLastID=nID;
		m_pGUI->Exit();
	}

	return FALSE;
}

void CMenuScene::Render()
{
	m_pGUI->Render();
}

File: GameMaker/Core/Menu Scene.lua
require "Common Const";

function Construct(this)
	LastID=0;
	Tex=NULL;
	Sound=NULL;
	Font=NULL;
	GUI=NULL;
	Spr=NULL;
end

function Deconstruct(this)
end

function Enter(this, Param)
	Tex=Texture_Load("cursor.png");
	Sound=Effect_Load("menu.wav");
	Font=Font_Create(40, 0, 0, false, "Î¢ÈíÑÅºÚ");

	Spr=Sprite_Create(Tex, 0, 0, 32, 32);

	GUI=GUI_Create();
	GUI_AddCtrl(GUI, Control_Create(1, "Menu Item.lua", {Font=Font, Sound=Sound, x=400, y=225, Delay=0.1, Title="Start"}));
	GUI_AddCtrl(GUI, Control_Create(2, "Menu Item.lua", {Font=Font, Sound=Sound, x=400, y=275, Delay=0.2, Title="Resume"}));
	GUI_AddCtrl(GUI, Control_Create(3, "Menu Item.lua", {Font=Font, Sound=Sound, x=400, y=325, Delay=0.3, Title="Exit"}));
	GUI_SetNavMode(GUI, Bit_or(NAV_UPDOWN, NAV_CYCLED));
	GUI_SetCursor(GUI, Spr);
	GUI_SetFocus(GUI, 1);
	GUI_Enter(GUI);

	LastID=0;

	return true;
end

function Exit(this)
	GUI_Free(GUI);
	Sprite_Free(Spr);
	Font_Free(Font);
	Audio_Free(Sound);
	Texture_Free(Tex);
end

function Update(this, Delta)
	if Input_GetKey()==VK_ESCAPE then
		LastID=3;
		GUI_Exit(GUI);
	end

	local ID=GUI_Update(GUI, Delta);

	if ID==-1 then
		if LastID==1 then
			SceneManager_Switch(Scene_GetManager(this), SCENE_STARS, {Mode=ENTER_START});
			return false;
		elseif LastID==2 then
			SceneManager_Switch(Scene_GetManager(this), SCENE_STARS, {Mode=ENTER_RESUME});
			return false;
		elseif LastID==3 then return true; end
	elseif ID>0 then
		LastID=ID;
		GUI_Exit(GUI);
	end

	return false;
end

function Render(this)
	GUI_Render(GUI);
end
